channeldb/migration25/migration.go

```diff
@@ -0,0 +1,119 @@
+)
+
+// MigrateInitialBalances patches the two new fields, InitialLocalBalance and
+// InitialRemoteBalance, for all the open channels. It does so by reading the
+// revocation log at height 0 to learn the initial balances and then updates
+// the channel's info.
+// The channel info is saved in the nested bucket which is accessible via
+// nodePub:chainHash:chanPoint. If any of the sub-buckets turns out to be nil,
+// we will log the error and continue to process the rest.
+func MigrateInitialBalances(tx kvdb.RwTx) error {
+	log.Infof("Migrating initial local and remote balances...")
+
+	openChanBucket := tx.ReadWriteBucket(openChannelBucket)
+
+	// If no bucket is found, we can exit early.
+	if openChanBucket == nil {
+		return nil
+	}
+
+	// Iterate the root bucket.
+	err := openChanBucket.ForEach(func(nodePub, v []byte) error {
+		// Ensure that this is a key the same size as a pubkey, and
+		// also that it leads directly to a bucket.
+		if len(nodePub) != 33 || v != nil {
+			return nil
+		}
+
+		// Read the sub-bucket level 1.
+		nodeChanBucket := openChanBucket.NestedReadWriteBucket(nodePub)
+		if nodeChanBucket == nil {
+			log.Errorf("no bucket for node %x", nodePub)
+			return nil
+		}
+
+		// Iterate the bucket.
+		return nodeChanBucket.ForEach(func(chainHash, _ []byte) error {
+			// Read the sub-bucket level 2.
+			chainBucket := nodeChanBucket.NestedReadWriteBucket(
+				chainHash,
+			)
+			if chainBucket == nil {
+				log.Errorf("unable to read bucket for chain=%x",
+					chainHash)
+				return nil
+			}
+
+			// Iterate the bucket.
+			return chainBucket.ForEach(func(cp, _ []byte) error {
+				// Read the sub-bucket level 3.
+				chanBucket := chainBucket.NestedReadWriteBucket(
+					cp,
+				)
+				if chanBucket == nil {
+					log.Errorf("unable to read bucket for ",
+						"chanPoint=%x", cp)
+					return nil
+				}
+
+				return migrateBalances(chanBucket)
+			})
+		})
+	})
+
+	return err
+}
+
+// migrateBalances queries the revocation log at height 0 to find the initial
+// balances and save them to the channel info.
+func migrateBalances(chanBucket kvdb.RwBucket) error {
+	c := &OpenChannel{}
+
+	// Get the old channel info.
+	if err := fetchChanInfo(chanBucket, c, true); err != nil {
+		return fmt.Errorf("unable to fetch chan info: %v", err)
+	}
+
+	// Fetch the channel commitments, which are useful for freshly open
+	// channels as they don't have any revocation logs and their current
+	// commitments reflect the initial balances.
+	if err := FetchChanCommitments(chanBucket, c); err != nil {
+		return fmt.Errorf("unable to fetch chan commits: %v", err)
+	}
+
+	// Get the initial balances.
+	localAmt, remoteAmt, err := c.balancesAtHeight(chanBucket, 0)
+	if err != nil {
+		return fmt.Errorf("unable to get initial balances: %v", err)
+	}
+
+	c.InitialLocalBalance = localAmt
+	c.InitialRemoteBalance = remoteAmt
+
+	// Update the channel info.
+	if err := putChanInfo(chanBucket, c, false); err != nil {
```

Comment:
IIRC, we can't modify a bucket while we're doing a `ForEach` range over it. Typically, we do one pass to collect the set of keys we need to modify, then another pass to actually write the keys.

Reply:
Just recalled you'd reminded me before...will change.